cardano-cli/src/Cardano/CLI/Ping.hs

```diff
@@ -0,0 +1,582 @@
+keepAliveDone v | v >= NodeToNodeVersionV7 minBound minBound =
+    CBOR.toLazyByteString $
+         CBOR.encodeListLen 1
+      <> CBOR.encodeWord 2
+keepAliveDone _ =
+    CBOR.toLazyByteString $
+      CBOR.encodeWord 2
+
+
+handshakeReqEnc :: [NodeVersion] -> CBOR.Encoding
+handshakeReqEnc [] = error "null version list"
+handshakeReqEnc versions =
+      CBOR.encodeListLen 2
+  <> CBOR.encodeWord 0
+  <> CBOR.encodeMapLen (fromIntegral $ L.length versions)
+  <> mconcat [ encodeVersion v
+              | v <- versions
+              ]
+  where
+    encodeVersion :: NodeVersion -> CBOR.Encoding
+    encodeVersion (NodeToClientVersionV9 magic) =
+          CBOR.encodeWord (9 `setBit` nodeToClientVersionBit)
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToClientVersionV10 magic) =
+          CBOR.encodeWord (10 `setBit` nodeToClientVersionBit)
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToClientVersionV11 magic) =
+          CBOR.encodeWord (11 `setBit` nodeToClientVersionBit)
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToClientVersionV12 magic) =
+          CBOR.encodeWord (12 `setBit` nodeToClientVersionBit)
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToClientVersionV13 magic) =
+          CBOR.encodeWord (13 `setBit` nodeToClientVersionBit)
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToClientVersionV14 magic) =
+          CBOR.encodeWord (14 `setBit` nodeToClientVersionBit)
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV1 magic) =
+          CBOR.encodeWord 1
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV2 magic) =
+          CBOR.encodeWord 2
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV3 magic) =
+          CBOR.encodeWord 3
+      <>  CBOR.encodeInt (fromIntegral magic)
+    encodeVersion (NodeToNodeVersionV4 magic mode) = encodeWithMode 4 magic mode
+    encodeVersion (NodeToNodeVersionV5 magic mode) = encodeWithMode 5 magic mode
+    encodeVersion (NodeToNodeVersionV6 magic mode) = encodeWithMode 6 magic mode
+    encodeVersion (NodeToNodeVersionV7 magic mode) = encodeWithMode 7 magic mode
+    encodeVersion (NodeToNodeVersionV8 magic mode) = encodeWithMode 8 magic mode
+    encodeVersion (NodeToNodeVersionV9 magic mode) = encodeWithMode 9 magic mode
+    encodeVersion (NodeToNodeVersionV10 magic mode) = encodeWithMode 10 magic mode
+
+
+    encodeWithMode :: Word -> Word32 -> Bool -> CBOR.Encoding
+    encodeWithMode vn magic mode =
+          CBOR.encodeWord vn
+      <>  CBOR.encodeListLen 2
+      <>  CBOR.encodeInt (fromIntegral magic)
+      <>  CBOR.encodeBool mode
+
+handshakeReq :: [NodeVersion] -> ByteString
+handshakeReq []       = BL.empty
+handshakeReq versions = CBOR.toLazyByteString $ handshakeReqEnc versions
+
+data HandshakeFailure
```

Comment:
Why is all of the handshake related stuff being implemented here? Does this not exist in consensus?

Reply:
The code is not yet exported by `ouroboros-network`.  The code that should be exported by `ouroboros-network` has been moved to `Cardano.CLI.Ping.Lib`.

After we're happy with the code, I will move it to `ouroboros-network`.  We may merge this code before the move, then perform the move, then wait for the next release, then update the code to use the exported code.

Reply:
The library has been moved to `ouroboros-network`.